Userland/Utilities/userdel.cpp

```diff
@@ -103,20 +141,35 @@ int main(int argc, char** argv)
         rc = 6;
     }
 
-    if (rc == 0 && chmod(temp_filename, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+    if (rc == 0 && chmod(temp_passwd, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+        perror("chmod");
+        rc = 1;
+    }
+
+    if (rc == 0 && chmod(temp_shadow, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
         perror("chmod");
         rc = 1;
     }
 
-    if (rc == 0 && rename(temp_filename, "/etc/passwd") < 0) {
+    if (rc == 0 && rename(temp_passwd, "/etc/passwd") < 0) {
         perror("failed to rename the temporary passwd file");
         rc = 1;
     }
 
+    if (rc == 0 && rename(temp_shadow, "/etc/shadow") < 0) {
+        perror("failed to rename the temporary shadow file");
+        rc = 1;
+    }
+
     if (rc) {
-        if (unlink(temp_filename) < 0) {
+        if (unlink(temp_passwd) < 0) {
             perror("unlink");
         }
+
+        if (unlink(temp_shadow) < 0) {
+            perror("unlink");
+        }
+
         return rc;
     }
```

Comment:
To make unlinking the temp files in various error cases easier I suggest we use an `ArmedScopeGuard` which we would then disarm here.

Something like this at the top:

```cpp
auto unlink_temp_files = [&] {
    if (unlink(temp_passwd) < 0)
        perror("unlink");
    if (unlink(temp_shadow) < 0)
        perror("unlink");
};
ArmedScopeGuard unlink_temp_files_guard = [&] {
    unlink_temp_files();
};
```

That way, instead of storing `rc` everywhere and subsequently checking for `rc == 0`, we can just `return` in all error cases and the scope guard will take care of it.

Instead of this `if (rc) {}` block we would then do:

```cpp
unlink_temp_files();
unlink_temp_files_guard.disarm();
```

(as far as I can tell we're currently failing to unlink the files in a non-error `rc == 0` case?)

Reply:
Thanks for the feedback! I agree, adding the arm guard makes it much cleaner. Went ahead and added it as a separate commit